### ISSUE TITLE
Optional dependency on transformers

### ISSUE THREAD

**Issue by DuarteMRAlves** [linkedin/Liger-Kernel#258]
### 🚀 The feature, motivation and pitch

I want to utilize the liger-kernel fused operations on a codebase but do not need the requirement for transformers. However, when I import the liger_kernel.transformers package, it also imports the transformers library.

### Alternatives

I can install transformers as a dependency, but it should not be required if it is not used.

### Additional context

I understand that one of the goals of this project is the seamless drop in replacement for transformer models, but I believe it should be possible to import the operations without such a requirement.

**Comment by yundai424** [comment]
if i understand correctly the only place that pulls `transformers` in when trying to import `liger_kernel.transformers` is [here](https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/transformers/__init__.py#L1). It's possible to make it an optional import (i.e. only import `auto_model` when `transformers` is available). Wdyt? @shimizust @ByronHsu

**Comment by shimizust** [comment]
Good catch, we should wrap that in optional import. Will take this

**Comment by shimizust** [comment]
@yundai424 I think the monkeypatch functions are also pulling in transformers via the imports at top as well (e.g. `from liger_kernel.transformers.model.mistral import lce_forward as mistral_lce_forward`). Will move these down into the individual functions

**Comment by shimizust** [comment]
Thinking about it more, @DuarteMRAlves what exactly are you trying to import? The raw ops can be imported from `liger_kernel.ops` and don't require transformers. Everything under `liger_kernel.transformers` is meant for replacing modules in specific models added in the transformers library. So wondering what you're using here if also not using transformers

**Comment by DuarteMRAlves** [comment]
Thank you for the replies.

I submitted this issue because I was using the fused linear cross entropy loss, which is relatively model agnostic. As far as I saw it, there would be three levels of the api:

* The triton ops which already do not require transformers;
* The nn modules which wrap the ops to be used in any model (and would not require transformers);
* The transformer patches which are very transformer specific.

In the meantime, I have also adopted the swiglu, rms norm and rope in my codebase. Here, I agree with your point that these follow the transformers api very closely, namely in their config, and I ended up directly importing the ops package.

For these later modules, although I believe they could be useful to codebases which do not expose transformers, their coupling to the transformers implementation makes it much harder to do so.

Possibly, a middle ground would be a note on how liger kernel should be used outside of transformers, because the fused linear cross entropy example in the readme "guides" the user towards importing the transformers package.
